drivers: flash: Add driver for AT45 compatible SPI flash chips

anangl opened the pull request:
Add a driver that can handle several instances of AT45 family chips, which are enabled by specifying DT nodes for them with the "compatible" property set to "atmel,at45" and other required properties like JEDEC ID, chip capacity, block and page size etc. configured accordingly.

The driver is only capable of using "power of 2" binary page sizes in those chips and at initialization configures them to work in that mode (unless it is already done).

A sample application is also provided, intended to present capabilities of the driver and to serve as a reference of how to make use of the driver.

anangl committed:
drivers: flash: Add driver for AT45 compatible SPI flash chips

Add a driver that can handle several instances of AT45 family chips,
which are enabled by specifying DT nodes for them with the "compatible"
property set to "atmel,at45" and other required properties like JEDEC
ID, chip capacity, block and page size etc. configured accordingly.

The driver is only capable of using "power of 2" binary page sizes in
those chips and at initialization configures them to work in that mode
(unless it is already done).

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
samples: drivers: Add sample that shows how to use AT45 flash driver

This commit adds a sample application intended to present capabilities
of the flash driver for AT45 family chips and to serve as a reference
of how to make use of that driver.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

nordic-auko commented on drivers/flash/spi_flash_at45d.c:
```diff
@@ -0,0 +1,695 @@
+		return 0;
+	}
+
+	u8_t const conf_binary_page_size[] = {
```
Should these values be listed among the CMD_ defines?

nordic-auko commented on drivers/flash/spi_flash_at45d.c:
```diff
@@ -0,0 +1,695 @@
+static int perform_chip_erase(struct device *dev)
+{
+	int err;
+	u8_t const chip_erase_cmd[] = {
```
Should these values be listed among the CMD_ defines?

nordic-auko reviewed (commented):
Looks good, only minor comments. Tested on AT45DB321, erasing/writing/reading entire flash memory range. Did not measure current consumption for DPD and UDPD yet.

anangl commented:
@nordic-auko Thanks for the review. I addressed your comments.

I also removed a part of the sample that performed an extra erasing. This part was primarily used for testing the driver and it did not bring much value to the sample.

Other changes applied:
- added `sample.yaml` so that the sample gets built by sanitycheck
- added to `nrf9160_pca10090.overlay` deletion of UART's RX pin assignment, so that the receiver is not activated at all and does not consume power unnecessarily

nvlsianpu commented on drivers/flash/Kconfig.at45d:
```diff
@@ -0,0 +1,35 @@
+
+if SPI_FLASH_AT45D
+
+config SPI_FLASH_AT45D_USE_READ_MODIFY_WRITE
```
Why this is no default y? This is my expectation to any flash drive and I believe most of others expectation as well..

nvlsianpu replied:
@anangl?

anangl replied:
Oh, right, I modified the sample accordingly but forgot to add this actual `default y` here. Sorry for that. Fixed now.

nvlsianpu reviewed (commented):
I appreciate smart `erase` algorithm.
just a few coding-style nips

nvlsianpu commented on samples/drivers/spi_flash_at45d/sample.yaml:
```diff
@@ -0,0 +1,6 @@
+tests:
+  sample.drivers.spi.flash.at45d:
+    tags: spi flash
+    platform_whitelist: nrf9160_pca10090
```
can harness be added?
This will help to verify sample using sanity-check.
Example of such harness: https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/drivers/soc_flash_nrf/sample.yaml#L7-L23

pabigot commented on drivers/flash/spi_flash_at45d.c:
```diff
@@ -0,0 +1,692 @@
+	}
+
+	u8_t const op_and_addr[] = {
+		[0] = CMD_READ,
```
I don't think designated initializers add any value when it's a contiguous gapless sequence using literal ordinals.  A standard initializer list would be less confusing.

pabigot commented on dts/bindings/mtd/adesto,at45d.yaml:
```diff
@@ -0,0 +1,59 @@
+# Copyright (c) 2020 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+description: Adesto AT45D family DataFlash SPI Memory
```
The compatible for this driver should be `atmel,at45` as it is in Linux.  Adesto is just a vendor conforming to that API.  The D part of the model identifier does not affect the definition.

pabigot commented on dts/bindings/mtd/adesto,at45d.yaml:
```diff
@@ -0,0 +1,59 @@
+    required: false
+    default: 0
+    description: |
+      Time, in microseconds, needed by the chip to enter the Deep Power-Down
```
Delays should be specified in nanoseconds.

pabigot reviewed (changes requested):
Initial comments which require some rework:

* Use `atmel,at45` as the compatible and `at45` for driver name (the `d` does not appear to be used to distinguish this; see also at24 and at25.  (Technically it could be `dataflash` as it is in Linux, but that's an Atmel registered trademark when capitalized, and I'm not sure that at45 is the only command set the Linux driver supports.)
* Use nanoseconds for devicetree properties that measure time for things like command delays.
* Avoid designated initializers for sequential array initialization.

I also agree on replacing the magic numbers.

I'll look more soon but wanted to get these changes queued.

anangl commented:
@nvlsianpu @pabigot I addressed your comments, i.e:
- used "atmel,at45" as the compatible and at45 instead of at45d in the driver and the corresponding sample names
- got rid of pointed out magic numbers
- removed uses of designated initializers
- switched to nanoseconds in introduced devicetree properties

anangl commented:
I'll do one more update of this PR.  I'll rebase it on the latest master and update the overlay file with the new name of the board, and I'll use the new devicetree API in the driver.

anangl commented:
Rebased on master.

anangl commented:
Updated the overlay file name and converted the driver to the new devicetree API.

pabigot commented on samples/hello_world/nrf52840dk_nrf52840.overlay:
```diff
@@ -0,0 +1,3 @@
+&uart0 {
```
Does this belong here?  It doesn't look like it belongs here....

anangl replied:
Of course not. I grabbed this change accidentally. Shame on me.

pabigot commented on drivers/flash/spi_flash_at45.c:
```diff
@@ -0,0 +1,685 @@
+	u32_t chip_size;
+	u16_t block_size;
+	u16_t page_size;
+	u16_t t_enter_dpd; /* in microseconds */
```
update comment with corrected units.

anangl replied:
But these time values are in microseconds. They are converted from nanoseconds at [initialization](https://github.com/zephyrproject-rtos/zephyr/blob/869dfbb3f49d890546019112d881f8062de2e833/drivers/flash/spi_flash_at45.c#L665) of this structure.

pabigot replied:
OK, fine then, thanks for the clarification.

pabigot commented on drivers/flash/spi_flash_at45.c:
```diff
@@ -0,0 +1,685 @@
+		LOG_ERR("SPI transaction failed with code: %d/%u",
+			err, __LINE__);
+	}
+
```
I think there should be a `*status = sys_le16_to_cpu(*status)` here on success to put the status register into host byte order.

anangl replied:
Good point. Added.

pabigot reviewed (changes requested):
Three small change (units in comment, byte order of status word, plus removing the hello world sample change).  I built and ran this but I don't have the modified hardware so I can't independently verify functionality.
